### PR TITLE
#1 PR adds conjured items enhancement

### PR DESCRIPTION
**Pull request by ikarabulut**
We promise this PR is not as large as it looks. It covers enhancement #1 which also had tasks #6 and #15 attached to it as well. 

**Minor details left out of this PR**
- Packages: currently the classes are floating around in a giant pond. We plan to add packages to organize these but want to keep that as a separate Pr.

Our approach with this implementation was to isolate as much behavior as we could in order to open the system up for easier maintainability and extensibility. To do this we implemented the strategy pattern to implement an `Updatable` interface which acts as the base behavior of an item. The protocol for this interface is to have an `update()` method. Since each item has a different way of updating its `sellIn` and `quality` those behaviors were extracted into new classes that implement the `Updateable` interface. within each Updateable class implemented `update` method you set the behavior of how much the `value` passed (quality or sellIn) will increase or decrease depending on the class name.

These behaviors were then composed into the new `DefaultItem` class hierarchy which has a `qualityBehavior` `sellInBehavior` instance variables which create the `Updateable` behavior. 

Within the `DefaultItem` hierarchy the protocol is to `age` the item. Only each item will know how they age based on `sellIn`, and those implementation details are left in the private methods you'll see. 

TL;DR
- Followed Strategy Pattern
- Increase or decrease behavior extracted into a new interface: How quality and sellIn go up or down is based on the objects that implement `Updateable`
- New item behaviors are now in the `DefaultItem` class hierarchy

### CODE REVIEW

File: src/main/java/com/gildedrose/AgedBrie.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package com.gildedrose;
+
+public class AgedBrie extends DefaultItem {
```
Comment (mrfishball) [inline src/main/java/com/gildedrose/AgedBrie.java]: Questions: I'm curious about the use of `extends`(inheritance) over `implements`(composition). By using inheritance, we're coupling this class with `DefaultItem` class. This means that any changes to `DefaultItem` will affect or alter the behaviors of `AgedBrie` in ways that we might not expect.

Comment (ikarabulut) [reply src/main/java/com/gildedrose/AgedBrie.java]: That is a good question. And this idea flowed over into this new iteration of how we implemented the strategy pattern.

Our goal with this was to set the default behaviors for all items, which is to decrease quality by 1, and decrease sellIn by 1. And there are a few overlaps where we will still use that default like Conjured where it takes the default and doubles it. And we will run into this same use-case with the Red Wine requirement in the pipeline.

A potential abstraction that we have not yet uncovered is what determines how quickly something degrades which seems to be based on sellIn. We don't have enough knowledge of the business domain based on the limited number of products. We are hoping as more products flow into the program we will be able to identify a pattern and then create class hierarchies based on that pattern. i.e "cheese will double in quality when past the sellIn" we would create a superclass of `cheese` then subclass any other types of cheeses that might have slight variances

We could remove the `extends` from AgedBrie and Backstage since they don't rely on the default update behaviors. And then see if the abstraction above we can use to create a new Item type hierarchy.

We'd love to hear some more thoughts and ideas on this.

Comment (riccjohn) [reply src/main/java/com/gildedrose/AgedBrie.java]: Be careful of trying to optimize your code now for future tickets. I know you can see the upcoming tickets, so it's not like you're 100% trying to predict the future, but try to write the code that fits the ticket you're currently on and don't optimize too early. That said, I agree that it's easy to reach. for inheritance, but it's often not the right approach. I'm still going through the PR, so hopefully I can give some more concrete feedback / ideas around it, but that's just my initial thought.

File: src/main/java/com/gildedrose/DecreaseBy2.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.gildedrose;
+
+public class DecreaseBy2 implements Updatable {
```
Comment (elijahsullivan) [inline src/main/java/com/gildedrose/DecreaseBy2.java]: This is along the same lines as what was mentioned by @mrfishball and @riccjohn in this thread https://github.com/jramnani/GildedRoseJavaLLC/pull/16/files#r921567623

I'm wondering if at the item level (e.g. "Conjured", "AgedBrie", etc.) you're reaching for inheritance where composition might be more appropriate. Similarly here–I'm wondering if composition, and even the classes, are necessary or if this arithmetic could just live in the implementations of `updateQuality` and `updateSellIn` in each of the respective `Item` classes.

Phrased another way–I'm not sure these classes are buying us much:

`src/main/java/com/gildedrose/DecreaseToZero.java`
`src/main/java/com/gildedrose/DefaultDecrease.java`
`src/main/java/com/gildedrose/DefaultIncrease.java`
`src/main/java/com/gildedrose/IncreaseBy2.java`
`src/main/java/com/gildedrose/IncreaseBy3.java`

Comment (riccjohn) [inline src/main/java/com/gildedrose/DecreaseBy2.java]: Suggestion/Blocker: I think this abstraction goes too far. Seeing classes that only augment one property, and do it in such a specific, single-use way, is a code-smell and tells me at least this part of the pattern is not the correct solution. For example, if we kept adding items that had their quality updated by a different amount, you'd have to keep adding new classes here.

I'm not sure yet if a larger structural change is needed or not, but some things on my mind are:
- The `DecreaseBy2` and `DecreaseToZero` both have an `update` method, but it's used in different ways. In one, you give it the value to subtract from, and in the other, you give it the value, but then do nothing with it and just return 0. That's a few more code-smells.
  - Implementations of a method shouldn't act that differently from one another
  - You don't want to hand a class/method an argument that goes unused. That tells me there's something off with this pattern / it's not the right abstraction.
- Seeing that all these classes that implement `Updatable` need to know about the idea of `quality` makes me think that the internals of the `Item` class are getting too spread out in your system. The idea of quality now jumps from Item > GildedRose > AgedBrie (or any other similar class) > Updatable > DefaultIncrease (or any other class that implements Updatable)

**Review by riccjohn — CHANGES_REQUESTED** [review]

File: src/main/java/com/gildedrose/DecreaseBy2.java (hunk 2 above)
Comment (riccjohn) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: Yup, I agree. I left a similar comment below, but the idea is the same. I also agree with Eli's idea of just keeping that logic within the respective Item classes for now.

Comment (elijahsullivan) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: +1

Comment (mrfishball) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: +1

File: src/main/java/com/gildedrose/GildedRose.java [hunk 3]
```diff
@@ -8,55 +8,30 @@ public GildedRose(Item[] items) {
...
-                        }
-                    } else {
-                        items[i].quality = items[i].quality - items[i].quality;
-                    }
-                } else {
-                    if (items[i].quality < 50) {
-                        items[i].quality = items[i].quality + 1;
-                    }
-                }
+        for (Item item : items) {
```
Comment (mrfishball) [inline src/main/java/com/gildedrose/GildedRose.java:11]: Question: I'm curious if all of this can just be just:
```Java
for (Item item : items) {
  item.age();
}
```
I know that we can't modify the `Item` class at all but you could wrap Item in another class.

Right now this class isn't open for an extension nor closed for modification because we have to come in and add another case every time we add a new item.

File: src/main/java/com/gildedrose/AgedBrie.java (hunk 1 above)
Comment (mrfishball) [reply src/main/java/com/gildedrose/AgedBrie.java:3]: Question: Since `Aged Brie` and `Backstage passes` age the same way, they would both extend the same class?

Comment (mrfishball) [reply src/main/java/com/gildedrose/AgedBrie.java]: Question: Is the idea to have `DefaultItem`, `legendaryItem` and then each item class will extend those based on how they age?

Comment (ikarabulut) [reply src/main/java/com/gildedrose/AgedBrie.java]: > Be careful of trying to optimize your code now for future tickets. I know you can see the upcoming tickets, so it's not like you're 100% trying to predict the future, but try to write the code that fits the ticket you're currently on and don't optimize too early. That said, I agree that it's easy to reach. for inheritance, but it's often not the right approach. I'm still going through the PR, so hopefully I can give some more concrete feedback / ideas around it, but that's just my initial thought.

Is this _really_ optimizing too much? When we decided to go down this route my initial reaction was "this is going to be a lot of small classes" but then my head went to, "this early on, it seems smarter to break this apart as much as possible to have a solid foundation to build upon" Since we really don't know what this software is going to become and we inherited this giant block of code I thought it would be best to isolate as much behavior as we can and something as simple as how things increment up and down seemed pretty safe that the behavior of how things appreciate or depreciate will stay pretty consistent but the implementation details of that is what will be most volatile, so with how it is currently set up, the program will be very flexible to those changes.

I would say that a miss was how we implemented the Items. We isolated the behavior of how things update but looking back may have been a little lazy with the items. But, the items are the biggest unknown here since we only have 5-6 in the program and pipeline so who knows what they will look like moving forward. Taking that into consideration maybe there would be a better way to refactor the items to keep them flexible but this early on, with how things are broken apart, it seems easier to put things back together to make changes rather than having to break them apart down the road based on a missed abstraction.

@riccjohn Im basing a lot of this methodology on 99 bottles where Sandi broke the problem down into very isolated pieces almost right off the bat based on the first requirement being introduced.

Comment (ikarabulut) [reply src/main/java/com/gildedrose/AgedBrie.java]: > Question: Is the idea to have DefaultItem, legendaryItem and then each item class will extend those based on how they age?

@mrfishball This is my opinion here without consulting with Jon. But that is a possibility. So far we only have one legendary item so who knows what the next one will look like, the behaviors might be totally different from the first one. That is my thought process around this at least

File: src/main/java/com/gildedrose/DecreaseBy2.java (hunk 2 above)
Comment (ikarabulut) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: > I'm wondering if at the item level (e.g. "Conjured", "AgedBrie", etc.) you're reaching for inheritance where composition might be more appropriate.

I agree here. Composition would have been a better call here, idk why my mind was pushing me toward inheritance so much with this abstraction, if we can even call it that.

> if this arithmetic could just live in the implementations of updateQuality and updateSellIn in each of the respective Item classes.

Ok, now I am starting to understand the idea of too much abstraction here. I guess, yes, this would be a good option. It would save us time to get the feature implemented without overoptimizing. Its like if we stopped at like 60% of what we were doing. 

I guess my question to the group here is what is the tradeoff to the approach of not abstracting to the point where we got to? I still don't see much harm in extracting those behaviors to isolate them as an `Update` abstraction. It allows for maximum flexibility with the one behavior whos implementation details are so far the most volatile in the program based on what we have in front of us. It did not really take much longer for us to get to that point. Yes, the classes we created are quite shallow and as the program expands there is a potential to a future developer getting lost in the details but I feel like with proper package setup it would be easy to locate the right class for what you are trying to do.

Comment (ikarabulut) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: >The DecreaseBy2 and DecreaseToZero both have an update method, but it's used in different ways. In one, you give it the value to subtract from, and in the other, you give it the value, but then do nothing with it and just return 0. That's a few more code-smells.

When this was implemented my spidey senses did activate for a split second that something was off. I just didn't think to much into it. But this is good feedback, and something I can take moving forward regarding the implementations of a method.

As Im typing this out I feel like perhaps the abstraction could have been `sellInBehavior` and `qualityBehavior` since those are really what we are looking at. The fact that it is `updateable` doesn't really give the next developer much. But if it were `sellInBehavior` and then the `update` method will describe the update behavior of sellIn. (if we were to stay on the path of strategy pattern)

Comment (ikarabulut) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: Also with this being a blocker. What is the correct approach here? Do I ask, what does unblocking this look like for you?

Comment (mrfishball) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: @ikarabulut One of the downsides of this approach is that you'll need a class for every variant, for example, what happens if you need to decrease by 5? or Increase to 20? The other downside is calling `DecreaseBy*` adds a layer of indirection to something that can just be `sellIn - 5` of `quality - 5` without having a clear benefit of either improving reusability or clarity of the code. In the end, you get potentially a growing number of classes that all do the same thing and you have to maintain them over time.

Comment (riccjohn) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: +1 to @mrfishball. I think this leads to confusing code for other devs to read and it'll become difficult to maintain over time. Even you both as the authors of the code will have a tough time reading it after you haven't touched that part of the codebase for a few weeks or months. If our company merges with another company and suddenly we need to sell 50 new items, this pattern would be really cumbersome.

You mentioned not knowing what's coming next. I 100% agree with that, but since I don't know what types of items are coming next, I'd only do the refactoring that's totally necessary, and wait for patterns to emerge later on as you get new items in the store.

Comment (riccjohn) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: I think just doing what Steven suggested above:

> if this arithmetic could just live in the implementations of updateQuality and updateSellIn in each of the respective Item classes.

I'd remove all the `DecreaseBy2` -type classes and put that logic back in the respective item classes.

File: src/main/java/com/gildedrose/AgedBrie.java (hunk 1 above)
Comment (mrfishball) [reply src/main/java/com/gildedrose/AgedBrie.java]: I see. I would be careful when defining relationships this way because then you're opening up the space to have many many sub-classes that might not make sense. Instead of saying this item is a special item and special items do X, Y or Z, perhaps look at it like these are all `Item` and item A can do X, item B can do Y.

File: src/main/java/com/gildedrose/GildedRose.java (hunk 3 above)
Comment (jonmanahan) [reply src/main/java/com/gildedrose/GildedRose.java:11]: We were thinking about doing this near the beginning but felt as though it was kind of working around the requirement the client set in a way that could upset the goblin and we really didn't want to get one shotted, that being said we are not opposed to the idea in general. Our other thought was going the route of creating a hashmap wrapper (called handlers or something) that implemented the builder pattern, whose creation and population exists on the app level (so devs could easily access and interact with it).
The hashmap would contain a string key (the names of the items) and the value would be the corresponding way each items quality/sellIn modification are handled via a `Consumer Interface`. A `getOrDefault()` could be used so that if the name does not exist in the hashmap wrapper, the defaultitem consumer interface function would be called.

Here is a photo to help understand my blob of words:
image
And an example `Consumer Interface` Method (Aged Brie):
image


However I think your idea/comment is much simpler and better if we are allowed to implement it! What are everyones thoughts on this?
Also if anything was unclear in my explanation please let me know 😄 thank you for help!

File: src/main/java/com/gildedrose/DecreaseBy2.java (hunk 2 above)
Comment (jonmanahan) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: This makes sense to us, we removed the extends `DefaultItem` from every special class (except for `Conjured`) and removed all of the `DecreaseBy2` - type classes and put that logic back in the respective item classes.

The reason we kept the `Conjured extends DefaultItem` is because the functionality of the quality degradation of a `ConjuredItem` is inherently linked/coupled with that of the DefaultItem. This being "`Conjured` items degrade in Quality twice as fast as normal items".

This blocker solution is addressed from [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/787edac2b69be3565984ac07747d058edd313cba) to [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/b4a7a025d8519c2f1355aff41986085493192ee6) (small commits so the whole solution spanned across multiple)

Comment (jonmanahan) [reply src/main/java/com/gildedrose/DecreaseBy2.java]: The removal of the extends and `Updateable` classes in each item is addressed from [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/787edac2b69be3565984ac07747d058edd313cba) to [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/b4a7a025d8519c2f1355aff41986085493192ee6) (small commits so the whole solution spanned across multiple)

File: src/main/java/com/gildedrose/AgedBrie.java (hunk 1 above)
Comment (jonmanahan) [reply src/main/java/com/gildedrose/AgedBrie.java]: The removal of the extends and `Updateable` classes in each item is addressed from [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/787edac2b69be3565984ac07747d058edd313cba) to [this commit](https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/b4a7a025d8519c2f1355aff41986085493192ee6) (small commits so the whole solution spanned across multiple)

**Review by riccjohn — APPROVED** [review]

File: src/main/java/com/gildedrose/Conjured.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public void age(Item item) {
+        item.quality = updateQuality(item.quality, item.sellIn);
+        item.sellIn = updateSellIn(item.sellIn);
+    }
+
+    private int updateQuality(int quality, int sellIn) {
+        if (sellIn <= 0) {
+            conjuredQualityDegradation = super.qualityDegradation * 2;
```
Comment (riccjohn) [inline src/main/java/com/gildedrose/Conjured.java]: Question: Does this `conjuredQualityDegredation` value need to be set here if it's already set in the constructor on line 7?

File: src/main/java/com/gildedrose/GildedRose.java (hunk 3 above)
Comment (mrfishball) [reply src/main/java/com/gildedrose/GildedRose.java:11]: **Suggestion:** I realized my previous comment is confusing. Let me clarify. When I mentioned the wrapper class, I meant something like:
```java
class AgedBrie extends Item {
  ...
}
```

and then in `GildedRose` I wonder if we can just do this:
```java
  ...
  for (Item item : items) {
    item.age();
  }
```

I think an `Ageable` interface might be more appropriate for composition. Like:
```java
interface Ageable {
  void age(int sellIn, int quality);
}
```
 This goes back to one of the comments I left about using inheritance. We must be careful when modeling the relationship between the sub-class and the super class.

File: src/main/java/com/gildedrose/Conjured.java (hunk 4 above)
Comment (ikarabulut) [reply src/main/java/com/gildedrose/Conjured.java]: This actually exposed a missing edge case in our test harness that Conjured doubles after the sellIn date so if default goes to 2 when passed sellIn then Conjured should be 4. Plus the syntax needed to be `... = conjuredQualityDegredation * 2` This was addressed just now in this commit https://github.com/jramnani/GildedRoseJavaLLC/pull/16/commits/b88740b67b9b2c72b7d494e7e2ece559cfef73bd

File: src/main/java/com/gildedrose/GildedRose.java (hunk 3 above)
Comment (ikarabulut) [reply src/main/java/com/gildedrose/GildedRose.java:11]: Can we talk through this some time on Monday? Im also confused by this approach because the way I see it we would still need the `Item` class to implement this `ageable` interface? I might be following you but it might be easier to talk through.

**Review by riccjohn — APPROVED** [review]

**Comment by ikarabulut** [comment]
Merging into `trunk`

File: src/main/java/com/gildedrose/GildedRose.java (hunk 3 above)
Comment (mrfishball) [reply src/main/java/com/gildedrose/GildedRose.java:11]: For the record, Jon, Ish, and I had a call to clarify this comment. Essentially, you can use `extends` and `implements` at the same time on `AgedBrie` without modifying `Item` if you so choose:
```java
class AgedBrie extends Item implements Ageable { ... }
```
